Add support for Assimp skins

LogicAndTrick opened the pull request:
Sorry for the spam, this one should be good now...

I'm working through my changes to support HL1 models and trying to split the work up into multiple PRs (as per #4284). This change is for a completely different feature than the other PR, sorry about that... But I think this one makes more sense to apply first because the animation stuff can be applied on top of it much more cleanly.

The bulk of this change is to refactor the way the `AssimpParser` works to remove the 4 class fields that just store temporary data for the purposes of creating the model. The purpose of this was to add support for multiple skins in Assimp models.

- `AssimpComputedMeshData` was added to hold a mesh and its VBO data
- `getMeshIndex` performs a simple lookup of mesh -> mesh index as Assimp does not have a built in way for this
- Loaded models are changed from one surface with all textures, to multiple surfaces with one texture each (plus skins, if present)
    - `AssimpMeshWithTransforms` was added to pass through mesh information
- Vertex data is computed when building the surface meshes, rather than populating the `m_vertices` field beforehand

I realise I'm still including multiple changes in one PR but I'm hoping the changes here are easier to review. The inline diff looks a bit messy but it makes a bit more sense with the side-by-side diff.

FGD change to test skins:
```
@PointClass base(Monster, TalkMonster)
  size(-16 -16 0, 16 16 72)
  model({ "path": "models/scientist.mdl", "skin": skin }) // <---- add the "skin" part here
= monster_scientist : "Scared Scientist" 
[ ... ]
```

I'm hoping this level of change is ok, I can break this change down a bit more (remove the skins loading and just have the refactoring) but it would mean I'd be doing extra work to split out the changes just for them to be thrown away once I log the next follow-up PR. The skins functionality is simply loading an array of textures rather than just one texture, so most of it is just the refactoring.

LogicAndTrick commented:
I wanted to give this one another shot, so here's my attempt at doing the same thing in smaller commits :)

LogicAndTrick commented:
Anything I can do to make reviewing this a bit easier? I did the smaller commits with the idea of them being squash merged at the end, but now I'm thinking that maybe its better if it was more like 2-3 commits that get merged/rebased normally instead?

kduske commented:
Hey, the problem isn't the PR as such, just that I have very little time for TB at the moment. My highest priority is to fix one outstanding severe issue with protected properties, so if I do have time for TB, I work on that. 

Sorry that it takes so long!

LogicAndTrick commented:
All good! Just wanted to make sure the problem wasn't on my side :)

kduske commented on common/src/IO/AssimpParser.cpp:
```diff
@@ -226,24 +224,32 @@ std::unique_ptr<Assets::EntityModel> AssimpParser::doInitializeModel(
+    const std::string meshName = mesh->mName.data;
+
+    // load skins for this surface
+    const size_t meshMaterialIndex = mesh->mMaterialIndex;
```
Use auto.

kduske commented on common/src/IO/AssimpParser.cpp:
```diff
@@ -226,24 +224,32 @@ std::unique_ptr<Assets::EntityModel> AssimpParser::doInitializeModel(
+
+    // load skins for this surface
+    const size_t meshMaterialIndex = mesh->mMaterialIndex;
+    auto textures = std::vector<Assets::Texture>{};
```
You can shorten this to `surface.setSkins({processMaterial(...)});`

kduske commented on common/src/IO/AssimpParser.cpp:
```diff
@@ -256,61 +262,70 @@ void AssimpParser::loadSceneFrame(
-        return v.attr;
-      });
+    // build the mesh faces as triangles
+    const size_t numTriangles = mesh.m_mesh->mNumFaces;
```
auto, auto, auto ;-)

kduske reviewed (changes requested):
All in all this looks good, but I have left a bunch of comments. I have two more things I'd like to change here: All the static member functions of `AssimpParser` should now be turned into free functions in the anonymous namespace. And any function that takes a pointer that can never be null (such as `aiScene*`) should take that value by reference instead.

Finally - would it be feasible to add a test that verifies that this change is working?

LogicAndTrick commented on common/src/IO/AssimpParser.cpp:
```diff
@@ -226,24 +224,32 @@ std::unique_ptr<Assets::EntityModel> AssimpParser::doInitializeModel(
+
+    // an assimp mesh will only ever have one material, but a material can have multiple
+    // alternatives (this is how assimp handles skins)
+    const std::string meshName = mesh->mName.data;
```
Turns out `meshName` isn't even used, so I just removed it entirely.

LogicAndTrick commented on common/src/IO/AssimpParser.cpp:
```diff
@@ -404,9 +407,9 @@ std::optional<Assets::Texture> loadFallbackTexture(const FileSystem& fs)
+  const aiScene& scene, size_t i, Logger& logger) const
 {
-  for (unsigned int i = 0; i < scene.mNumMaterials; i++)
   {
```
I think it's just due to the way I structured the commits - I removed the braces in a later commit to try and keep the diff clean with fewer whitespace changes. Probably ended up being more confusing than helpful in the end!

kduske replied:
I see, nevermind then. Remember that you can set github's diff viewer to ignore whitespace changes, so that's not a big concern anyway ;-)

LogicAndTrick commented on common/src/IO/AssimpParser.cpp:
```diff
@@ -425,39 +428,40 @@ void AssimpParser::processMaterials(const aiScene& scene, Logger& logger)
+      {
+        return std::move(*fallbackTexture);
+      }
+      else
```
Was changed in a later commit to add to a vector instead of returning instantly.

LogicAndTrick commented on common/src/IO/AssimpParser.cpp:
```diff
@@ -425,39 +428,40 @@ void AssimpParser::processMaterials(const aiScene& scene, Logger& logger)
-        m_textures.push_back(loadTextureFromFileSystem(filePath, m_fs, logger));
+        return loadTextureFromFileSystem(filePath, m_fs, logger);
       }
       else if (texture->mHeight != 0)
```
Same as previous comment

LogicAndTrick commented on common/src/IO/AssimpParser.cpp:
```diff
@@ -425,39 +428,40 @@ void AssimpParser::processMaterials(const aiScene& scene, Logger& logger)
+        return loadUncompressedEmbeddedTexture(
+          texture->pcData, texture->mFilename.C_Str(), texture->mWidth, texture->mHeight);
       }
       else
```
Same

LogicAndTrick commented on common/src/IO/AssimpParser.cpp:
```diff
@@ -423,7 +423,7 @@ Assets::Texture loadUncompressedEmbeddedTexture(
 
 Assets::Texture loadCompressedEmbeddedTexture(
-  std::string name,
+  const std::string& name,
```
Just to check I understand this right, do you mean the code should instead be this?

```cpp
Assets::Texture loadCompressedEmbeddedTexture(
  std::string name,
  ...)
{
  return readFreeImageTextureFromMemory(
           std::move(name), ...)
    ...;
}
```

kduske replied:
Yes, exactly!

LogicAndTrick commented on common/src/IO/AssimpParser.cpp:
```diff
@@ -225,14 +222,17 @@ void AssimpParser::loadSceneFrame(
   {
-    bounds.add(std::begin(m_positions), std::end(m_positions));
+    bounds.add(
+      std::begin(m_vertices), std::end(m_vertices), [](const Assets::EntityModelVertex& v) {
```
This ended up being changed in a later commit.

kduske commented:
Will you let me know when this is ready for another look?

LogicAndTrick commented:
Ok, a little sooner than I expected, but I've added the test case and fixed up that last minor code change. It's ready for another review now!

I made the cube.mdl file and released it under CC0 [here](https://github.com/LogicAndTrick/halflife-test-assets).

kduske commented:
superseded by https://github.com/TrenchBroom/TrenchBroom/pull/4395
